plugins/usbdmx/NodleU1.cpp

```diff
@@ -0,0 +1,528 @@
+        m_adaptor(adaptor) {
+  }
+
+ private:
+  LibUsbAdaptor* const m_adaptor;
+
+  bool ReceiveBuffer(libusb_device_handle *handle,
+                     DmxBuffer *buffer,
+                     bool *buffer_updated);
+  bool ReadDataChunk(libusb_device_handle *handle,
+                     uint8_t *usb_data);
+};
+
+bool NodleU1ThreadedReceiver::ReceiveBuffer(libusb_device_handle *handle,
+                                            DmxBuffer *buffer,
+                                            bool *buffer_updated) {
+  unsigned char usb_data[DATABLOCK_SIZE];
+
+  if (ReadDataChunk(handle, usb_data)) {
+    if (usb_data[0] < 16) {
+      uint16_t startOff = usb_data[0] * 32;
+      for (int i = 0; i < 32; i++) {
+        buffer->SetChannel(startOff + i, usb_data[i + 1]);
+      }
+      *buffer_updated = true;
+    }
+  }
+
+  return true;
+}
+
+bool NodleU1ThreadedReceiver::ReadDataChunk(libusb_device_handle *handle,
+                                            uint8_t *usb_data) {
+  int transferred;
+  int ret = m_adaptor->InterruptTransfer(
+      handle, READ_ENDPOINT, reinterpret_cast<unsigned char*>(usb_data),
+      DATABLOCK_SIZE, &transferred, URB_TIMEOUT_MS);
+  if (ret && ret != LIBUSB_ERROR_TIMEOUT) {
+    OLA_WARN << "InterruptTransfer(): " << m_adaptor->ErrorCodeToString(ret)
+             << ", transferred " << transferred << " / " << DATABLOCK_SIZE;
+  }
+  return ret == 0;
+}
+
+// SynchronousNodleU1
+// -----------------------------------------------------------------------------
+
+SynchronousNodleU1::SynchronousNodleU1(
+    LibUsbAdaptor *adaptor,
+    libusb_device *usb_device,
+    const string &serial,
+    unsigned int mode)
+    : NodleU1(adaptor, serial, mode),
+      m_usb_device(usb_device) {
+}
+
+bool SynchronousNodleU1::Init() {
+  libusb_device_handle *usb_handle = OpenNodleU1Widget(
+      m_adaptor, m_usb_device);
+
+  if (!usb_handle) {
+    return false;
+  }
+
+  SetInterfaceMode(m_adaptor, usb_handle, m_mode);
+
+  if (m_mode & 2) {  // output port active
+    std::auto_ptr<NodleU1ThreadedSender> sender(
+        new NodleU1ThreadedSender(m_adaptor, m_usb_device, usb_handle));
+    if (!sender->Start()) {
+      return false;
+    }
+    m_sender.reset(sender.release());
+  }
+
+  if (m_mode & 4) {  // input port active
+    std::auto_ptr<NodleU1ThreadedReceiver> receiver(
+        new NodleU1ThreadedReceiver(m_adaptor, m_usb_device, usb_handle));
+    if (!receiver->Start()) {
+      return false;
+    }
+    m_receiver.reset(receiver.release());
+  }
+
+  return true;
+}
+
+bool SynchronousNodleU1::SendDMX(const DmxBuffer &buffer) {
+  return m_sender.get() ? m_sender->SendDMX(buffer) : false;
+}
+
+void SynchronousNodleU1::SetDmxCallback(Callback0<void> *callback) {
+  if (m_receiver.get()) {
+    m_receiver->SetReceiveCallback(callback);
+  }
+}
+
+const DmxBuffer &SynchronousNodleU1::GetDmxInBuffer() const {
+  return m_receiver->GetDmxInBuffer();
```

Comment:
But this would require me to return another DmxBuffer...
So it left it like that because it should not get called anyway when the receiver was not initialized.